Userland/Applications/FTPServer/FTPServerClient.cpp

```diff
@@ -0,0 +1,1169 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "FTPServerClient.h"
+#include <AK/JsonArray.h>
+#include <AK/JsonObject.h>
+#include <AK/MappedFile.h>
+#include <LibCore/DateTime.h>
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <LibCore/FileStream.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#undef FTP_DEBUG
+#define FTP_DEBUG 1
+
+int FTPServerClient::m_next_data_port = 4000;
+
+FTPServerClient::FTPServerClient(int id, RefPtr<Core::TCPSocket> socket)
+    : m_id(id)
+    , m_socket(move(socket))
+{
+    m_socket->on_ready_to_read = [this] { drain_socket(); };
+}
+
+static String get_current_ip()
```

Comment:
This approach seems error prone. Presumably you want to select the IP address for the server adapter interface associated with the `m_socket` client connection socket instead?

If that's not possible for some reason, this should be marked as a TODO / FIXME.

Reply:
Method removed, as the user of it now calls the socket for the address instead